Slave motors AND endstops

Moonlight63 opened the pull request:
I'll keep this short and sweet. I don't full expect this to be merged upstream, but just want everyone to know it's here if they need it.

I'm building a CNC using smoothie and have 2 motors for both my Y and Z axis. I wanted to add the ability to use endstops to square the gantry. Marlin has this functionality. 

This change allows each stepper to define a slave, and each endstop may define a slave that controls the slave stepper. The slave endstops only trigger on homing actions, the master switch may still be used as a hard limit. This type of configuration only makes sense on cartesian frames and is completely optional. To use the slave stepper, simply define the slave step, direction, and enable pins:
```
alpha_slave_step_pin
alpha_slave_dir_pin
alpha_slave_en_pin
```
Using slave endstops requires using the new syntax. This would technically work with A,B,C axis if you had a board with that many outputs. I'm using external drivers, so any output pins will work. All that is needed is:
`endstop.minx.slave_pin`

That's pretty much all there is to it. You can slave the motor without using a slaved endstop, but in that case there really is no point. I would love to turn this into a self contained module, but I don't think that is possible.

CascadeAP committed:
Added motor and endstop pairing feature

CascadeAP committed:
Well that was emberessing

CascadeAP committed:
I don't feel as bad about these oversights

CascadeAP committed:
I clearly misunderstood how things work, but it works now.

arthurwolf commented:
Hey.

Thanks for the code!
I've seen somebody try to contribute this before, but as far as I can see
this is better implemented than the other time was.
I don't think this would get merged into mainstream v1, however, we have a
procedure for features that are useful to some but can't be merged: can you
please add a link to your branch to this page:
http://smoothieware.org/third-party-branches ?

And also ideally you should add a link to this at the bottom of the current
method to do this with Smoothie:
http://smoothieware.org/switch?s[]=multiple&s[]=homing#homing-a-multi-motor-axis

How hard would it be to make this support more than 2 endstops for an axis
( there are a few people with 4-motors/endstops Z axes ) ?

Cheers !



On Fri, Sep 27, 2019 at 11:19 AM Moonlight63 <notifications@github.com>
wrote:

> I'll keep this short and sweet. I don't full expect this to be merged
> upstream, but just want everyone to know it's here if they need it.
>
> I'm building a CNC using smoothie and have 2 motors for both my Y and Z
> axis. I wanted to add the ability to use endstops to square the gantry.
> Marlin has this functionality.
>
> This change allows each stepper to define a slave, and each endstop may
> define a slave that controls the slave stepper. The slave endstops only
> trigger on homing actions, the master switch may still be used as a hard
> limit. This type of configuration only makes sense on cartesian frames and
> is completely optional. To use the slave stepper, simply define the slave
> step, direction, and enable pins:
>
> alpha_slave_step_pin
> alpha_slave_dir_pin
> alpha_slave_en_pin
>
> Using slave endstops requires using the new syntax. This would technically
> work with A,B,C axis if you had a board with that many outputs. I'm using
> external drivers, so any output pins will work. All that is needed is:
> endstop.minx.slave_pin
>
> That's pretty much all there is to it. You can slave the motor without
> using a slaved endstop, but in that case there really is no point. I would
> love to turn this into a self contained module, but I don't think that is
> possible.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/Smoothieware/Smoothieware/pull/1431
> Commit Summary
>
>    - Added motor and endstop pairing feature
>    - Well that was emberessing
>    - I don't feel as bad about these oversights
>    - I clearly misunderstood how things work, but it works now.
>
> File Changes
>
>    - *M* src/libs/StepperMotor.cpp
>    <https://github.com/Smoothieware/Smoothieware/pull/1431/files#diff-0>
>    (20)
>    - *M* src/libs/StepperMotor.h
>    <https://github.com/Smoothieware/Smoothieware/pull/1431/files#diff-1>
>    (40)
>    - *M* src/modules/robot/Robot.cpp
>    <https://github.com/Smoothieware/Smoothieware/pull/1431/files#diff-2>
>    (19)
>    - *M* src/modules/tools/endstops/Endstops.cpp
>    <https://github.com/Smoothieware/Smoothieware/pull/1431/files#diff-3>
>    (82)
>    - *M* src/modules/tools/endstops/Endstops.h
>    <https://github.com/Smoothieware/Smoothieware/pull/1431/files#diff-4>
>    (4)
>
> Patch Links:
>
>    - https://github.com/Smoothieware/Smoothieware/pull/1431.patch
>    - https://github.com/Smoothieware/Smoothieware/pull/1431.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Smoothieware/Smoothieware/pull/1431?email_source=notifications&email_token=AAA2SFMZ2VDVNZLAAZMPVXLQLXF2ZA5CNFSM4I3D53JKYY3PNVWWK3TUL52HS4DFUVEXG43VMWVGG33NNVSW45C7NFSM4HODHLTA>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAA2SFP6VAZNDBR5XDWCQB3QLXF2ZANCNFSM4I3D53JA>
> .
>


-- 
Courage et bonne humeur.

wolfmanjm commented:
I would also not want to merge this as it adds valuable cycles to time critical areas of the code, for a feature I suspect very few will be able to use.

Moonlight63 commented:
Ok, I've added a link back here in both sections for anyone who needs it. I may also include a precompiled bin for anyone who wants this but is uncomfortable with ARM. I also just realized I called it slave motors and **encoders**. As cool as that would be, I was clearly on autopilot when typing that!

Theoretically this _could_ be expanded to work with more motors, and one idea I had for that was to refactor things to allow for an arbitrary number of slaves, such as:
```
alpha_slave1_step_pin
alpha_slave2_step_pin
alpha_slave3_step_pin
etc
```
This would be really cool, especially for doing something like 3 point bed leveling, but I think it would ultimately be pointless _for smoothie_. Smoothieboards only have 5 dedicated motor outputs with very little space for expandability. They aren't Duet boards. Making this type of change would require a massive change to the code base, which I was specifically trying to avoid here, but would also likely require a different board to be made altogether, likely with a bigger microcontroller. 
On that subject, making a bigger board where that type of feature would actually be useful should be relatively straight forward because, well, Mbed is awesome. I would love to try getting smoothie running on an STM32 just for the hell of it!
I _may_ try to build the arbitrary slaves feature at some point in the future just to see if I can, but I don't see it being all that useful for the time being due to board constraints. 

I tried to make sure that everything I changed wouldn't affect cycles if the feature was not being used, outside of checking if the feature is being used. Can you give any advise on how to reduce the time used in the time critical areas further? I suspect you are referring to the read_endstops function being called by the SlowTicker?

arthurwolf commented:
Thank you so much for the wiki edits. A bin ( you can also link it from the
third party page ) would be a good idea yes I think.

About config with multiple possible settings, the best way I think is to
use the same syntax we use with switch: alpha_slave.1.step_pin ( Smoothie
has a mechanism to break down config options written this way, see about
half of modules ).

About external drivers, there are thousands of users using Smoothie with
more than 5 drivers, it's as simple as wiring 2 pins to an external driver
...

About timing, I'm 99% certain Jim is talking about step generation.

On Sat, Sep 28, 2019 at 1:56 AM Moonlight63 <notifications@github.com>
wrote:

> Ok, I've added a link back here in both sections for anyone who needs it.
> I may also include a precompiled bin for anyone who wants this but is
> uncomfortable with ARM. I also just realized I called it slave motors and
> *encoders*. As cool as that would be, I was clearly on autopilot when
> typing that!
>
> Theoretically this *could* be expanded to work with more motors, and one
> idea I had for that was to refactor things to allow for an arbitrary number
> of slaves, such as:
>
> alpha_slave1_step_pin
> alpha_slave2_step_pin
> alpha_slave3_step_pin
> etc
>
> This would be really cool, especially for doing something like 3 point bed
> leveling, but I think it would ultimately be pointless *for smoothie*.
> Smoothieboards only have 5 dedicated motor outputs with very little space
> for expandability. They aren't Duet boards. Making this type of change
> would require a massive change to the code base, which I was specifically
> trying to avoid here, but would also likely require a different board to be
> made altogether, likely with a bigger microcontroller.
> On that subject, making a bigger board where that type of feature would
> actually be useful should be relatively straight forward because, well,
> Mbed is awesome. I would love to try getting smoothie running on an STM32
> just for the hell of it!
>
> I tried to make sure that everything I changed wouldn't affect cycles if
> the feature was not being used, outside of checking if the feature is being
> used. Can you give any advise on how to reduce the time used in the time
> critical areas further? I suspect you are referring to the read_endstops
> function being called by the SlowTicker?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/Smoothieware/Smoothieware/pull/1431?email_source=notifications&email_token=AAA2SFJ7ETRXIOLOOZRXYJ3QL2MSHA5CNFSM4I3D53JKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD72LFTI#issuecomment-536130253>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAA2SFJGRDASKMFZJPSQNR3QL2MSHANCNFSM4I3D53JA>
> .
>


-- 
Courage et bonne humeur.

sergiocantoro commented:
Hi @Moonlight63, i am trying to use your branch. Y slave motor motor works correctly while the endstop is not detected. These are my settings, are they wrong?

```
epsilon_current                                  1.8
beta_slave_step_pin                          2.8
beta_slave_dir_pin                             2.13!
beta_slave_en_pin                             4.29
endstop.miny.slave_pin                      1.27^! # YMIN pin on mks sbase 1.3
```

Moonlight63 commented:
That looks right to me, but it has been a while since I have used smoothieware. I would check your wiring and make sure signals are getting through. I think it was also set up to only work if the primary IO was also present. Meaning, you can't use slave Y motor and slave miny endstop if you aren't using the regular motor and endstop as well. They work in tandem. You don't have your full config here so I am just clarifying.

sergiocantoro commented:
Hi @Moonlight63 , this is my full config, as you can see i use two motors and two endstops on y axis

```
# Smoothieboard configuration file, see http://smoothieware.org/configuring-smoothie
# NOTE Lines must not exceed 132 characters, and '#' characters mean what follows is ignored
## Robot module configurations : general handling of movement G-codes and slicing into moves

# Basic motion configuration
default_feed_rate                            4000             # Default speed (mm/minute) for G1/G2/G3 moves
default_seek_rate                            4000             # Default speed (mm/minute) for G0 moves
mm_per_arc_segment                           0.0              # Fixed length for line segments that divide arcs, 0 to disable
#mm_per_line_segment                         5                # Cut lines into segments this size
mm_max_arc_error                             0.01             # The maximum error for line segments that divide arcs 0 to disable
                                                              # note it is invalid for both the above be 0
                                                              # if both are used, will use largest segment length based on radius

# Arm solution configuration : Cartesian robot. Translates mm positions into steplhper positions
# See http://smoothieware.org/stepper-motors
alpha_steps_per_mm                           80               # Steps per mm for alpha ( X ) stepper
beta_steps_per_mm                            80               # Steps per mm for beta ( Y ) stepper
gamma_steps_per_mm                           80               # Steps per mm for gamma ( Z ) stepper

# Planner module configuration : Look-ahead and acceleration configuration
# See http://smoothieware.org/motion-control
acceleration                                 3000             # Acceleration in mm/second/second.
#z_acceleration                              500              # Acceleration for Z only moves in mm/s^2, 0 uses acceleration which is the default. DO NOT SET ON A DELTA
junction_deviation                           0.05             # See http://smoothieware.org/motion-control#junction-deviation
#z_junction_deviation                        0.0              # For Z only moves, -1 uses junction_deviation, zero disables junction_deviation on z moves DO NOT SET ON A DELTA

# Cartesian axis speed limits
x_axis_max_speed                             30000            # Maximum speed in mm/min
y_axis_max_speed                             30000            # Maximum speed in mm/min
z_axis_max_speed                             300              # Maximum speed in mm/min

# Stepper module configuration
# Pins are defined as  ports, and pin numbers, appending "!" to the number will invert a pin
# See http://smoothieware.org/pin-configuration and http://smoothieware.org/pinout
alpha_step_pin                               2.0              # Pin for alpha stepper step signal
alpha_dir_pin                                0.5!             # Pin for alpha stepper direction, add '!' to reverse direction
alpha_en_pin                                 0.4              # Pin for alpha enable pin
alpha_current                                1.8              # X stepper motor current
alpha_max_rate                               30000.0          # Maximum rate in mm/min

beta_step_pin                                2.1              # Pin for beta stepper step signal
beta_dir_pin                                 0.11             # Pin for beta stepper direction, add '!' to reverse direction
beta_en_pin                                  0.10             # Pin for beta enable
beta_current                                 1.8              # Y stepper motor current
beta_max_rate                                30000.0          # Maxmimum rate in mm/min

gamma_step_pin                               2.2              # Pin for gamma stepper step signal
gamma_dir_pin                                0.20!            # Pin for gamma stepper direction, add '!' to reverse direction
gamma_en_pin                                 0.19             # Pin for gamma enable
gamma_current                                1.8              # Z stepper motor current
gamma_max_rate                               30000.0          # Maximum rate in mm/min

## Extruder module configuration
# See http://smoothieware.org/extruder
extruder.hotend.enable                          true          # Whether to activate the extruder module at all. All configuration is ignored if false
extruder.hotend.steps_per_mm                    140           # Steps per mm for extruder stepper
extruder.hotend.default_feed_rate               600           # Default rate ( mm/minute ) for moves where only the extruder moves
extruder.hotend.acceleration                    500           # Acceleration for the stepper motor mm/sec²
extruder.hotend.max_speed                       50            # Maximum speed in mm/s

extruder.hotend.step_pin                        2.3           # Pin for extruder step signal
extruder.hotend.dir_pin                         0.22          # Pin for extruder dir signal ( add '!' to reverse direction )
extruder.hotend.en_pin                          0.21          # Pin for extruder enable signal

# Extruder offset
#extruder.hotend.x_offset                        0            # X offset from origin in mm
#extruder.hotend.y_offset                        0            # Y offset from origin in mm
#extruder.hotend.z_offset                        0            # Z offset from origin in mm

# Firmware retract settings when using G10/G11, these are the defaults if not defined, must be defined for each extruder if not using the defaults
#extruder.hotend.retract_length                  3            # Retract length in mm
#extruder.hotend.retract_feedrate                45           # Retract feedrate in mm/sec
#extruder.hotend.retract_recover_length          0            # Additional length for recover
#extruder.hotend.retract_recover_feedrate        8            # Recover feedrate in mm/sec (should be less than retract feedrate)
#extruder.hotend.retract_zlift_length            0            # Z-lift on retract in mm, 0 disables
#extruder.hotend.retract_zlift_feedrate          6000         # Z-lift feedrate in mm/min (Note mm/min NOT mm/sec)

delta_current                                    1.5          # First extruder stepper motor current

# Second extruder module configuration
#extruder.hotend2.enable                         true         # Whether to activate the extruder module at all. All configuration is ignored if false
#extruder.hotend2.steps_per_mm                   140          # Steps per mm for extruder stepper
#extruder.hotend2.default_feed_rate              600          # Default rate ( mm/minute ) for moves where only the extruder moves
#extruder.hotend2.acceleration                   500          # Acceleration for the stepper motor, as of 0.6, arbitrary ratio
#extruder.hotend2.max_speed                      50           # mm/s

#extruder.hotend2.step_pin                       2.8          # Pin for extruder step signal
#extruder.hotend2.dir_pin                        2.13         # Pin for extruder dir signal ( add '!' to reverse direction )
#extruder.hotend2.en_pin                         4.29         # Pin for extruder enable signal

#extruder.hotend2.x_offset                       0            # x offset from origin in mm
#extruder.hotend2.y_offset                       25.0         # y offset from origin in mm
#extruder.hotend2.z_offset                       0            # z offset from origin in mm

#epsilon_current                                 1.5          # Second extruder stepper motor current


## Laser module configuration
# See http://smoothieware.org/laser
laser_module_enable                           true            # Whether to activate the laser module at all
laser_module_pwm_pin                          2.5             # This pin will be PWMed to control the laser.
                                                              # Only pins 2.0, 2.1, 2.2, 2.3, 2.4, 2.5, 1.18, 1.20, 1.21, 1.23, 1.24, 1.26, 3.25 and 3.26
                                                              # can be used since laser requires hardware PWM, see http://smoothieware.org/pinout
#laser_module_ttl_pin                      1.30            # This pin turns on when the laser turns on, and off when the laser turns off.
laser_module_maximum_power                    1.0             # This is the maximum duty cycle that will be applied to the laser
laser_module_minimum_power                    0.0             # This is a value just below the minimum duty cycle that keeps the laser
                                                              # active without actually burning.
laser_module_default_power                    0.8             # This is the default laser power that will be used for cuts if a power has not been specified.  The value is a scale between
                                                              # the maximum and minimum power levels specified above
laser_module_pwm_period      200      # This sets the pwm frequency as the period in microseconds

## Temperature control configuration
# See http://smoothieware.org/temperaturecontrol

# First hotend configuration
temperature_control.hotend.enable            true             # Whether to activate this ( "hotend" ) module at all.
temperature_control.hotend.thermistor_pin    0.23             # Pin for the thermistor to read
temperature_control.hotend.heater_pin        2.7              # Pin that controls the heater, set to nc if a readonly thermistor is being defined
temperature_control.hotend.thermistor        EPCOS100K        # See http://smoothieware.org/temperaturecontrol#toc5
#temperature_control.hotend.beta             4066             # Or set the beta value
temperature_control.hotend.set_m_code        104              # M-code to set the temperature for this module
temperature_control.hotend.set_and_wait_m_code 109            # M-code to set-and-wait for this module
temperature_control.hotend.designator        T                # Designator letter for this module
#temperature_control.hotend.max_temp         300              # Set maximum temperature - Will prevent heating above 300 by default
#temperature_control.hotend.min_temp         0                # Set minimum temperature - Will prevent heating below if set

# Safety control is enabled by default and can be overidden here, the values show the defaults
# See http://smoothieware.org/temperaturecontrol#runaway
#temperature_control.hotend.runaway_heating_timeout      900  # How long it can take to heat up, max is 2040 seconds.
#temperature_control.hotend.runaway_cooling_timeout        0  # How long it can take to cool down if temp is set lower, max is 2040 seconds
#temperature_control.hotend.runaway_range                20   # How far from the set temperature it can wander, max setting is 63°C

# PID configuration
# See http://smoothieware.org/temperaturecontrol#pid
#temperature_control.hotend.p_factor         13.7             # P ( proportional ) factor
#temperature_control.hotend.i_factor         0.097            # I ( integral ) factor
#temperature_control.hotend.d_factor         24               # D ( derivative ) factor

#temperature_control.hotend.max_pwm          64               # Max pwm, 64 is a good value if driving a 12v resistor with 24v.

# Second hotend configuration
#temperature_control.hotend2.enable            true           # Whether to activate this ( "hotend" ) module at all.
#temperature_control.hotend2.thermistor_pin    0.25           # Pin for the thermistor to read
#temperature_control.hotend2.heater_pin        1.23           # Pin that controls the heater
#temperature_control.hotend2.thermistor        EPCOS100K      # See http://smoothieware.org/temperaturecontrol#thermistor
##temperature_control.hotend2.beta             4066           # or set the beta value
#temperature_control.hotend2.set_m_code        104            # M-code to set the temperature for this module
#temperature_control.hotend2.set_and_wait_m_code 109          # M-code to set-and-wait for this module
#temperature_control.hotend2.designator        T1             # Designator letter for this module

#temperature_control.hotend2.p_factor          13.7           # P ( proportional ) factor
#temperature_control.hotend2.i_factor          0.097          # I ( integral ) factor
#temperature_control.hotend2.d_factor          24             # D ( derivative ) factor

#temperature_control.hotend2.max_pwm          64              # Max pwm, 64 is a good value if driving a 12v resistor with 24v.

temperature_control.bed.enable               true             # Whether to activate this ( "hotend" ) module at all.
temperature_control.bed.thermistor_pin       0.24             # Pin for the thermistor to read
temperature_control.bed.heater_pin           2.5              # Pin that controls the heater
temperature_control.bed.thermistor           Honeywell100K    # See http://smoothieware.org/temperaturecontrol#thermistor
#temperature_control.bed.beta                3974             # Or set the beta value
temperature_control.bed.set_m_code           140              # M-code to set the temperature for this module
temperature_control.bed.set_and_wait_m_code  190              # M-code to set-and-wait for this module
temperature_control.bed.designator           B                # Designator letter for this module

# Bang-bang ( simplified ) control
# See http://smoothieware.org/temperaturecontrol#bang-bang
#temperature_control.bed.bang_bang           false            # Set to true to use bang bang control rather than PID
#temperature_control.bed.hysteresis          2.0              # Set to the temperature in degrees C to use as hysteresis

## Switch modules
# See http://smoothieware.org/switch

# Switch module for fan control
switch.fan.enable                            true             # Enable this module
switch.fan.input_on_command                  M106             # Command that will turn this switch on
switch.fan.input_off_command                 M107             # Command that will turn this switch off
switch.fan.output_pin                        2.6              # Pin this module controls
switch.fan.output_type                       pwm              # PWM output settable with S parameter in the input_on_comand
#switch.fan.max_pwm                          255              # Set max pwm for the pin default is 255

#switch.misc.enable                          true             # Enable this module
#switch.misc.input_on_command                M42              # Command that will turn this switch on
#switch.misc.input_off_command               M43              # Command that will turn this switch off
#switch.misc.output_pin                      2.4              # Pin this module controls
#switch.misc.output_type                     digital          # Digital means this is just an on or off pin

#switch.bltouch.enable                    true         #
#switch.bltouch.output_pin                2.4          # sw pwm can use any digital pin
#switch.bltouch.input_on_command          M280         #
#switch.bltouch.input_off_command         M281         #
#switch.bltouch.output_type               swpwm        # sw pwm must be low frequency
#switch.bltouch.pwm_period_ms             20           # 50Hz
#switch.bltouch.startup_state             false        # start up in the stow state
#switch.bltouch.startup_value             7.43         # On boot it will go into stow mode, also set with the M281
#switch.bltouch.default_on_value          3.3          # if M280 is issued without S this is the value it will set it to


## Temperatureswitch
# See http://smoothieware.org/temperatureswitch
# Automatically toggle a switch at a specified temperature. Different ones of these may be defined to monitor different temperatures and switch different swithxes
# Useful to turn on a fan or water pump to cool the hotend
#temperatureswitch.hotend.enable              true            #
#temperatureswitch.hotend.designator          T               # first character of the temperature control designator to use as the temperature sensor to monitor
#temperatureswitch.hotend.switch              misc            # select which switch to use, matches the name of the defined switch
#temperatureswitch.hotend.threshold_temp      60.0            # temperature to turn on (if rising) or off the switch
#temperatureswitch.hotend.heatup_poll         15              # poll heatup at 15 sec intervals
#temperatureswitch.hotend.cooldown_poll       60              # poll cooldown at 60 sec intervals

## Endstops
# See http://smoothieware.org/endstops
endstops_enable                              true             # The endstop module is enabled by default and can be disabled here
#corexy_homing                               false            # Set to true if homing on a hbot or corexy
alpha_min_endstop                            1.24^!           # Pin to read min endstop, add a ! to invert if endstop is NO connected to ground
#alpha_max_endstop                           1.25^            # Pin to read max endstop, uncomment this and comment the above if using max endstops
alpha_homing_direction                       home_to_min      # Or set to home_to_max and set alpha_max and uncomment the alpha_max_endstop
alpha_min                                    10               # This gets loaded as the current position after homing when home_to_min is set
alpha_max                                    1000             # This gets loaded as the current position after homing when home_to_max is set
beta_min_endstop                             1.26^!           # Pin to read min endstop, add a ! to invert if endstop is NO connected to ground
#beta_max_endstop                            1.27^!           # Pin to read max endstop, uncomment this and comment the above if using max endstops
beta_homing_direction                        home_to_min      # Or set to home_to_max and set alpha_max and uncomment the alpha_max_endstop
beta_min                                     10               # This gets loaded as the current position after homing when home_to_min is set
beta_max                                     1000             # This gets loaded as the current position after homing when home_to_max is set
#gamma_min_endstop                           1.28^!           # Pin to read min endstop, add a ! to invert if endstop is NO connected to ground
#gamma_max_endstop                           1.29^            # Pin to read max endstop, uncomment this and comment the above if using max endstops
gamma_homing_direction                       home_to_min      # Or set to home_to_max and set alpha_max and uncomment the alpha_max_endstop
gamma_min                                    0                # This gets loaded as the current position after homing when home_to_min is set
gamma_max                                    700              # This gets loaded as the current position after homing when home_to_max is set

alpha_max_travel                             1000             # Max travel in mm for alpha/X axis when homing
beta_max_travel                              1000             # Max travel in mm for beta/Y axis when homing
gamma_max_travel                             700              # Max travel in mm for gamma/Z axis when homing

# Optional enable limit switches, actions will stop if any enabled limit switch is triggered
alpha_limit_enable                          true              # Set to true to enable X min and max limit switches
beta_limit_enable                           true              # Set to true to enable Y min and max limit switches
gamma_limit_enable                          false             # Set to true to enable Z min and max limit switches

# Endstops home at their fast feedrate first, then once the endstop is found they home again at their slow feedrate for accuracy
alpha_fast_homing_rate_mm_s                  50               # Alpha/X fast homing feedrate in mm/second
alpha_slow_homing_rate_mm_s                  25               # Alpha/X slow homing feedrate in mm/second
beta_fast_homing_rate_mm_s                   50               # Beta/Y  fast homing feedrate in mm/second
beta_slow_homing_rate_mm_s                   25               # Beta/Y  slow homing feedrate in mm/second
gamma_fast_homing_rate_mm_s                  4                # Gamma/Z fast homing feedrate in mm/second
gamma_slow_homing_rate_mm_s                  2                # Gamma/Z slow homing feedrate in mm/second

alpha_homing_retract_mm                      10               # Distance to retract from the endstop after it is hit for alpha/X
beta_homing_retract_mm                       10               # Distance to retract from the endstop after it is hit for beta/Y
gamma_homing_retract_mm                      1                # Distance to retract from the endstop after it is hit for gamma/Z

# Optional order in which axis will home, default is they all home at the same time,
# If this is set it will force each axis to home one at a time in the specified order
homing_order                                 XYZ              # X axis followed by Y then Z last
#move_to_origin_after_home                   false            # Move XY to 0,0 after homing
#endstop_debounce_count                       100              # Uncomment if you get noise on your endstops, default is 100
#endstop_debounce_ms                          1                # Uncomment if you get noise on your endstops, default is 1 millisecond debounce
#home_z_first                                true             # Uncomment and set to true to home the Z first, otherwise Z homes after XY

# End of endstop config
# Delete the above endstop section and uncomment next line and copy and edit Snippets/abc-endstop.config file to enable endstops for ABC axis
#include abc-endstop.config

## Z-probe
# See http://smoothieware.org/zprobe
zprobe.enable                                false           # Set to true to enable a zprobe
zprobe.probe_pin                             1.28!^          # Pin probe is attached to, if NC remove the !
zprobe.slow_feedrate                         5               # Mm/sec probe feed rate
#zprobe.debounce_ms                          1               # Set if noisy
zprobe.fast_feedrate                         100             # Move feedrate mm/sec
zprobe.probe_height                          5               # How much above bed to start probe
#gamma_min_endstop                           nc              # Normally 1.28. Change to nc to prevent conflict,

# Levelling strategy
# Example for 3-point levelling strategy, see wiki documentation for other strategies
#leveling-strategy.three-point-leveling.enable         true        # a leveling strategy that probes three points to define a plane and keeps the Z parallel to that plane
#leveling-strategy.three-point-leveling.point1         100.0,0.0   # the first probe point (x,y) optional may be defined with M557
#leveling-strategy.three-point-leveling.point2         200.0,200.0 # the second probe point (x,y)
#leveling-strategy.three-point-leveling.point3         0.0,200.0   # the third probe point (x,y)
#leveling-strategy.three-point-leveling.home_first     true        # home the XY axis before probing
#leveling-strategy.three-point-leveling.tolerance      0.03        # the probe tolerance in mm, anything less that this will be ignored, default is 0.03mm
#leveling-strategy.three-point-leveling.probe_offsets  0,0,0       # the probe offsets from nozzle, must be x,y,z, default is no offset
#leveling-strategy.three-point-leveling.save_plane     false       # set to true to allow the bed plane to be saved with M500 default is false

## Panel
# See http://smoothieware.org/panel
# Please find your panel on the wiki and copy/paste the right configuration here
panel.enable                                 false             # Set to true to enable the panel code

# Example for reprap discount GLCD
# on glcd EXP1 is to left and EXP2 is to right, pin 1 is bottom left, pin 2 is top left etc.
# +5v is EXP1 pin 10, Gnd is EXP1 pin 9
#panel.lcd                                   reprap_discount_glcd     #
#panel.spi_channel                           0                 # SPI channel to use  ; GLCD EXP1 Pins 3,5 (MOSI, SCLK)
#panel.spi_cs_pin                            0.16              # SPI chip select     ; GLCD EXP1 Pin 4
#panel.encoder_a_pin                         3.25!^            # Encoder pin         ; GLCD EXP2 Pin 3
#panel.encoder_b_pin                         3.26!^            # Encoder pin         ; GLCD EXP2 Pin 5
#panel.click_button_pin                      1.30!^            # Click button        ; GLCD EXP1 Pin 2
#panel.buzz_pin                              1.31              # Pin for buzzer      ; GLCD EXP1 Pin 1
#panel.back_button_pin                       2.11!^            # Back button         ; GLCD EXP2 Pin 8

panel.menu_offset                            0                 # Some panels will need 1 here

panel.alpha_jog_feedrate                     6000              # X jogging feedrate in mm/min
panel.beta_jog_feedrate                      6000              # Y jogging feedrate in mm/min
panel.gamma_jog_feedrate                     6000              # Z jogging feedrate in mm/min

panel.hotend_temperature                     185               # Temp to set hotend when preheat is selected
panel.bed_temperature                        60                # Temp to set bed when preheat is selected

## Custom menus : Example of a custom menu entry, which will show up in the Custom entry.
# NOTE _ gets converted to space in the menu and commands, | is used to separate multiple commands
custom_menu.power_on.enable                true              #
custom_menu.power_on.name                  Power_on          #
custom_menu.power_on.command               M80               #

custom_menu.power_off.enable               true              #
custom_menu.power_off.name                 Power_off         #
custom_menu.power_off.command              M81               #


## Network settings
# See http://smoothieware.org/network
network.enable                               false            # Enable the ethernet network services
network.webserver.enable                     true             # Enable the webserver
network.telnet.enable                        true             # Enable the telnet server
network.ip_address                           auto             # Use dhcp to get ip address
# Uncomment the 3 below to manually setup ip address
#network.ip_address                           192.168.3.222   # The IP address
#network.ip_mask                              255.255.255.0   # The ip mask
#network.ip_gateway                           192.168.3.1     # The gateway address
#network.mac_override                         xx.xx.xx.xx.xx.xx  # Override the mac address, only do this if you have a conflict

## System configuration
# Serial communications configuration ( baud rate defaults to 9600 if undefined )
# For communication over the UART port, *not* the USB/Serial port
uart0.baud_rate                              115200           # Baud rate for the default hardware ( UART ) serial port

second_usb_serial_enable                     false            # This enables a second USB serial port
#leds_disable                                true             # Disable using leds after config loaded
#play_led_disable                            true             # Disable the play led

# Kill button maybe assigned to a different pin, set to the onboard pin by default
# See http://smoothieware.org/killbutton
kill_button_enable                           true             # Set to true to enable a kill button
kill_button_pin                              2.12             # Kill button pin. default is same as pause button 2.12 (2.11 is another good choice)

#msd_disable                                 false            # Disable the MSD (USB SDCARD), see http://smoothieware.org/troubleshooting#disable-msd
#dfu_enable                                  false            # For linux developers, set to true to enable DFU

# Only needed on a smoothieboard
# See http://smoothieware.org/currentcontrol
currentcontrol_module_enable                 true             # Control stepper motor current via the configuration file

epsilon_current                              1.8
beta_slave_step_pin                          2.8
beta_slave_dir_pin                           2.13!
beta_slave_en_pin                            4.29
endstop.miny.slave_pin                       1.27^!

```
